core/src/kura.rs

```diff
@@ -255,6 +277,59 @@ impl Kura {
         Some(hash_array_guard[index])
     }
 
+    pub fn get_block_height_by_hash(&self, hash: &HashOf<VersionedCommittedBlock>) -> Option<u64> {
+        let hash_array_guard = self.block_hash_array.lock().expect("access hash array");
+
+        if *hash == Hash::zeroed().typed() {
+            None
+        } else {
+            match hash_array_guard.iter().position(|&x| x == *hash) {
+                None => None,
+                Some(index) => {
+                    Some(index as u64 + 1)
+                }
+            }
+        }
+    }
+
+    pub fn get_block_by_height(&self, block_height: u64) -> Option<Arc<VersionedCommittedBlock>> {
+        let mut arc_array_guard = self.block_arc_array.lock().expect("lock on arc array");
+        if block_height == 0 || block_height > arc_array_guard.len() as u64 {
+            return None;
+        }
+        let block_number : usize = (block_height - 1).try_into().expect("Failed to cast to u32.");
+
+        if let Some(block_arc) = arc_array_guard[block_number].as_ref() {
+            Some(Arc::clone(block_arc))
+        } else {
+            let block_store = self.block_store.lock().expect("lock on block_store");
+            let (block_start, block_len) = block_store.read_block_index(block_number as u64).expect("Failed to read block index from disk.");
+
+            let mut block_buf =
+                vec![0_u8; usize::try_from(block_len).expect("index_len didn't fit in 32-bits")];
+            block_store
+                .read_block_data(block_start, &mut block_buf)
+                .expect("Failed to read block data.");
+            let block = VersionedCommittedBlock::decode_versioned(&block_buf)
+                .expect("Failed to decode block");
+
+            let block_arc = Arc::new(block);
+            arc_array_guard[block_number] = Some(Arc::clone(&block_arc));
+            Some(block_arc)
+        }
```

Comment:
Technically, this is a `then_some` branch of the conditional. If you use `then_some` you get rid of the two `Some` instances, plus the `return None` line.